companion.py

```diff
@@ -128,6 +125,57 @@ def check_modules_ships(self) -> None:
             self.data['lastStarport']['ships'] = {'shipyard_list': {}, 'unavailable_list': []}
 
 
+class CAPIDataEncoder(json.JSONEncoder):
+    """Allow for json dumping via specified encoder."""
+
+    def default(self, o):
+        """Tell JSON encoder that we're actually just a dict."""
+        return o.__dict__
+
+
+class CAPIDataRawEndpoint:
+    """Last received CAPI response for a specific endpoint."""
+
+    def __init__(self, raw_data: str, query_time: datetime.datetime):
+        self.query_time = query_time
+        self.raw_data = raw_data
+        # TODO: Maybe requests.response status ?
+
+
+class CAPIDataRaw:
+    """The last obtained raw CAPI response for each endpoint."""
+
+    raw_data: Dict[str, CAPIDataRawEndpoint] = {}
+
+    def record_endpoint(
+            self, endpoint: str,
+            raw_data: str,
+            query_time: datetime.datetime
+    ):
+        """Record the latest raw data for the given endpoint."""
+        self.raw_data[endpoint] = CAPIDataRawEndpoint(raw_data, query_time)
+
+    def __str__(self):
+        """Return a more readable string form of the data."""
+        capi_data_str = '{'
+        for e in self.raw_data.keys():
+            capi_data_str += f'"{e}":\n{{\n\t"query_time": "{self.raw_data[e].query_time}",\n\t' \
+                             f'"raw_data": {self.raw_data[e].raw_data}\n}},\n\n'
+
+        capi_data_str = capi_data_str.removesuffix(',\n\n')
+        capi_data_str += '\n\n}'
+
+        return capi_data_str
```

Comment:
This isn't working.

1. The {} weren't encompassing the correct thing, but after fixing that...
2. the v.raw_data is a *string*, and that means when json.dumps()'d it all gets escaped, rather than being the raw data.  I do **not** want to json.loads() that only to json.dumps() it again, as it's meant to be the *raw* string (other than decode/encode passes).

I'll have a bit more of a play to see if I can at least use a generator.

Reply:
Nope, wanting the **raw** data is fundamentally incompatible with doing this.  I have changed my loop to `for k, v in self.raw_data.items()`, and as a consequence the references in the loop, though.